repo URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
commit ID: 

In [72]:
!python -m venv sklearn-env
!source sklearn-env/bin/activate
%pip install -U scikit-learn
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_absolute_error


Note: you may need to restart the kernel to use updated packages.


## Part 1. Loading the dataset

In [73]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
df = pd.read_csv('science_data_large.csv')
print(df.head(15))
summary_stats = df.describe()
print(summary_stats)
print(df.dtypes)


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500

## Part 2. Splitting the dataset

In [74]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.loc[:, df.columns != 'Size nm^3']
y = df['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [75]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Evaluate the model
# Select a sample data point from X_test
# Here, we will select the first sample as an example
sample_index = 0  # Change this index to select a different sample
sample_data = X_test.iloc[sample_index].values.reshape(1, -1)  # Reshape it into a 2D array

# Use the trained model to make a prediction
predicted_output = model.predict(sample_data)

print("Sample data point:", X_test.iloc[sample_index])  # Show the original unscaled sample
print("Predicted output for the sample data point:", predicted_output[0])
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
# Get the actual target value for the sample
actual_output = y_test.iloc[sample_index]

# Calculate the score (error)
mae = mean_absolute_error([actual_output], predicted_output)  # MAE for the sample
print("Actual output for the sample data point:", actual_output)
print("Mean Absolute Error (MAE) for this prediction:", mae)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
# Extract coefficients and intercept
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)

# Prepare to write the equation
feature_names = X.columns.tolist()

# Generate the LaTeX expression
equation = f"h(x) = {intercept:.4f}"  # Start with intercept
for coef, name in zip(coefficients, feature_names):
    equation += f" + ({coef:.4f}){name}"

# Display the LaTeX formatted equation
print(equation)

Sample data point: Temperature °C    100
Mols KCL          541
Name: 521, dtype: int64
Predicted output for the sample data point: 235911.19272259984
Actual output for the sample data point: 117762.3143
Mean Absolute Error (MAE) for this prediction: 118148.87842259984
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.4795834081
h(x) = -409391.4796 + (866.1464)Temperature °C + (1032.6951)Mols KCL


/Users/<redacted>/anaconda3/envs/cs383hw2/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Q: in your own words (markdown, not code) explain what the score means

A: The score is a measurement of prediction accuracy for regression models. It quantifies the average absolute difference between predicted values and actual values.

Sample equation: $E = mc^2$

$h(x) = -409391.4796 + (866.1464) \cdot \text{Temperature} \, (°C) + (1032.6951) \cdot \text{Mols KCL}$

## Part 4. Use Cross Validation

In [76]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
X_scaled = scaler.fit_transform(X)  # Fit and scale the data

# Use cross_val_score to evaluate the model with 5-fold cross-validation
scores = cross_val_score(model, X_scaled, y, cv=5, scoring='neg_mean_absolute_error')

# Convert scores from negative to positive MAE
mae_scores = -scores

# Print the results
print("Mean Absolute Error (MAE) scores for each fold:", mae_scores)
print("Mean MAE across all folds:", np.mean(mae_scores))
print("Standard Deviation of MAE:", np.std(mae_scores))

# Report on their finding and their significance
"""
The MAE scores range from 122501.9426136 to 130564.48002616, which indicates that
the model's performance might be sensitive to certain data point. 

On average, the predictions are off by approximately 125000.00 units from the actual 
target values. This is a relatively reasonable error range according to the observed 
data set. 

The standard deviation of MAE is approximately 3000.0, which suggests that the model 
is relatively stable.
"""

Mean Absolute Error (MAE) scores for each fold: [128113.39876727 130564.48002616 125295.84231418 122987.24100674
 122501.9426136 ]
Mean MAE across all folds: 125892.5809455914
Standard Deviation of MAE: 3067.643990719801


"\nThe MAE scores range from 122501.9426136 to 130564.48002616, which indicates that\nthe model's performance might be sensitive to certain data point. \n\nOn average, the predictions are off by approximately 125000.00 units from the actual \ntarget values. This is a relatively reasonable error range according to the observed \ndata set. \n\nThe standard deviation of MAE is approximately 3000.0, which suggests that the model \nis relatively stable.\n"

## Part 5. Using Polynomial Regression

In [83]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Create polynomial features of degree 2
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# Initialize the Linear Regression model
model_poly = LinearRegression()

# Train the model on the polynomial features
model_poly.fit(X_train_poly, y_train)

# Make predictions on the test set
y_pred_poly = model_poly.predict(X_test_poly)

# Calculate the Mean Absolute Error (MAE) on the test set
mae_poly = mean_absolute_error(y_test, y_pred_poly)

# Output the results
# print("Predicted values on test set (Polynomial Regression):", y_pred_poly)
# print("Actual values on test set:", y_test.to_numpy())
print("Mean Absolute Error (MAE) on test set (Polynomial Regression):", mae_poly)

# Report on the metrics and output the resultant equation as you did in Part 3.
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

print("Coefficients:", coefficients_poly)
print("Intercept:", intercept_poly)

# Prepare to write the polynomial equation
feature_names = poly.get_feature_names_out(X.columns)

# Generate the polynomial equation in LaTeX format
equation_poly = f"h(x) = {intercept_poly:.4f}"  # Start with intercept
for coef, name in zip(coefficients_poly, feature_names):
    equation_poly += f" + ({coef:.4f}){name}"

# Display the LaTeX formatted equation
print("LaTeX Equation for the Polynomial Model:")
print(equation_poly)

Mean Absolute Error (MAE) on test set (Polynomial Regression): 6.137037607004459e-05
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27196017e-07  1.26457259e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0478968508541584e-05
LaTeX Equation for the Polynomial Model:
h(x) = 0.0000 + (0.0000)1 + (12.0000)Temperature °C + (-0.0000)Mols KCL + (0.0000)Temperature °C^2 + (2.0000)Temperature °C Mols KCL + (0.0286)Mols KCL^2


$Simplified: h(x) = (12.0000)Temperature °C + (2.0000)Temperature °C Mols KCL + (0.0286)Mols KCL^2$